src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,411 @@
+            Collection<Slot> slots, Collection<Constraint> constraints,
+            QualifierHierarchy qualHierarchy, ProcessingEnvironment processingEnvironment,
+            Lattice lattice, Serializer<?, ?> defaultSerializer) {
+
+        BackEnd<?, ?> backEnd = null;
+
+        try {
+            Class<?> backEndClass = Class.forName(backEndType + "BackEnd");
+            Constructor<?> cons = backEndClass.getConstructor(Map.class, Collection.class,
+                    Collection.class, QualifierHierarchy.class, ProcessingEnvironment.class,
+                    Serializer.class, Lattice.class);
+            backEnd = (BackEnd<?, ?>) cons.newInstance(configuration, slots, constraints, qualHierarchy,
+                    processingEnvironment, defaultSerializer, lattice);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return backEnd;
+    }
+
+    /**
+     * This method is called when user doesn't separate constraints. Only one
+     * back end will be created.
+     * 
+     * @return an InferenceSolution for the given slots/constraints
+     */
+    protected InferenceSolution solve() {
+        solvingStart = System.currentTimeMillis();
+        Map<Integer, AnnotationMirror> result = realBackEnd.solve();
+        solvingEnd = System.currentTimeMillis();
+        StatisticRecorder.record(StatisticKey.OVERALL_NOGRAPH_SOLVING_TIME, (solvingEnd - solvingStart));
+        StatisticRecorder.record(StatisticKey.ANNOTATOIN_SIZE, (long) result.size());
+        PrintUtils.printResult(result);
+        return new DefaultInferenceSolution(result);
+    }
+
+    /**
+     * This method is called when user separates constraints, so that a list of
+     * back end is created for all components.
+     * 
+     * @param constraintGraph
+     * @param configuration
+     * @param slots
+     * @param constraints
+     * @param qualHierarchy
+     * @param processingEnvironment
+     * @param defaultSerializer
+     * @return an InferenceSolution for the given slots/constraints
+     */
+    protected InferenceSolution graphSolve(ConstraintGraph constraintGraph,
+            Map<String, String> configuration, Collection<Slot> slots,
+            Collection<Constraint> constraints, QualifierHierarchy qualHierarchy,
+            ProcessingEnvironment processingEnvironment, Serializer<?, ?> defaultSerializer) {
+
+        List<BackEnd<?, ?>> backEnds = new ArrayList<BackEnd<?, ?>>();
+        StatisticRecorder.record(StatisticKey.GRAPH_SIZE, (long) constraintGraph.getIndependentPath().size());
+
+        for (Set<Constraint> independentConstraints : constraintGraph.getIndependentPath()) {
+            backEnds.add(createBackEnd(backEndType, configuration, slots, independentConstraints,
+                    qualHierarchy, processingEnvironment, lattice, defaultSerializer));
+        }
+        // Clear constraint graph in order to save memory.
+        constraintGraph = null;
```

Comment:
Here we only set a local variable `constraintGraph` to `null`. But we still have a private field `constraintGraph` keep the reference to that graph. So I would guess here garbage collector doesn't destroy and re-collect the memory space of constraint graph.